0000-overhaul-qiskit-experiments.md

```diff
@@ -0,0 +1,638 @@
+        # 3. Run Analysis Executor, wait for results
+        # 4. Release lock of self.experiment_data
+        ...
+
+```
+
+It's noteworthy that one can also run experiment with transpiled circuits by directly using executor interface.
+See [Qiskit-Experiments/#1222](https://github.com/Qiskit-Extensions/qiskit-experiments/pull/1222) for background.
+
+```python
+exc = Executor()
+exc.job_executor.backend = some_backend
+exc.job_executor.add_circuits(transpiled_circuits)
+exc.analysis_executor.add_callback(T1Analysis())
+exc.run()
+
+exc.experiment_data.block_for_results()
+exc.experiment_data.save()
+```
+
+Note that [`BaseAnalysis.run`](https://github.com/Qiskit-Extensions/qiskit-experiments/blob/c66034c90dad73d705af25be7e9ed9617e7eb2ef/qiskit_experiments/framework/base_analysis.py#L118-L123) is not called in the new workflow.
+This method conventionally takes `ExperimentData` and mutates the input container through a callback.
+It internally wraps the `_run_analysis` method as a local function in which created analysis data are formatted and added to the experiment data.
+This local function is added to the input experiment data as a callback (very complicated!).
+By design, the analysis callback must NOT mutate the experiment data because the callback can run on a subprocess,
+and the mutated experiment data container doesn't update the original container residing in the main process.
+Therefore, the `BaseAnalysis.run` method is going to be deprecated, and instead we can add
+
+```python
+class BaseAnalysis:
+    ...
+
+    def __call__(
+        self, 
+        experiment_data: ExperimentData,
+    ) -> List[AnalysisResultData, Figure]:
+        return self._run_analysis(experiment_data)
+```
+
+to turn analysis instance into callable. This allows the executor to directly submit analysis instances.
+Eventually, we can also deprecate the `_run_analysis` method.
+Experiment developer will directly overwrite the call dunder method to provide the implementation of analysis.
+
+## Alternative Approaches
+In the proposed design, the executor is hidden and the experiment class behaves as the sole user interface.
+Alternatively, we can expose executor and remove `BaseExperiment.run` method.
+
+```python
+with Executor() as exc:
+    exp = MyExperiment((0,), **options)
+    exp_data = exc.run(exp)
+```
+
+On the one hand, this syntax has a high affinity with Runtime session, and we can completely delegate
+the role of circuit and analysis execution to the `Executor`. Experiment class becomes just a factory of experimental circuits.
+On the other hand, this introduces a breaking API change also to end users, and preclude us from capability of single liner execution with method chain.
+
+```python
+exp_data = MyExperiment(
+    (0,), backend=backend, **options
+).run().block_for_results()
+```
+
+If we have enough bandwidth to support end-user migration, alternative implementation is worth considering.
+
+
```

Comment:
I think we need more discussion around the alternative approach (i.e. getting rid of `BaseExperiment.run`), which indeed brings us closer to the runtime way of running. I don't know how acute is the single line of execution capability to our users. Personally I prefer 3 vertical calls instead of 3 horizontal (chained) calls; this way the code better captures the relevant steps, it's easier to debug if needed, profile etc.. But in any case we can keep `BaseExperiment.run` for compatibility (even after the deprecation period) in parallel of having executor-base flow supported.

Reply:
If I can design the pattern from scratch without considering the backward compatibility, I would propose something like this.

```python
with Executor(backend, shots=1000, block=True) as exp_data:
    qe.run(MyExperiment((0,), **options))  # get executor object from the contextvar
```

Context is very powerful, for example, I don't need explicit composite experiment classes:

```python
with Executor(backend, shots=1000, block=True) as exp_data:
    with qe.parallel():
        for q in qubits:
            qe.run(MyExperiment((q,), **options))  # get executor object from the contextvar
```

Reply:
That's nice but I would also consider a non-context based syntax, just not to enforce using `with` blocks for users who prefer the standard code layout for some reason

Reply:
If runtime sessions will become the default for most users, I think the context based syntax as the Q-E default is fine. I like the idea of keeping `BaseExperiment.run` for compatibility but promoting this context-based syntax as the standard pattern.

Reply:
I added more comment about extendability in [a99c946](https://github.com/Qiskit/RFCs/pull/47/commits/a99c94639c13fdbc853c09bd77bca0e3668318dc). Let's separately discuss execution pattern and focus on internal rework in this RFC. We need more user feedback to decide the interface.